src/hardware/rdtech-um/protocol.c

```diff
@@ -0,0 +1,281 @@
+/*
+ * This file is part of the libsigrok project.
+ *
+ * Copyright (C) 2018-2020 Andreas Sandberg <andreas@sandberg.pp.se>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+#include <glib.h>
+#include <libsigrok/libsigrok.h>
+#include "libsigrok-internal.h"
+#include "protocol.h"
+
+#define SERIAL_WRITE_TIMEOUT_MS 1
+
+#define UM_POLL_LEN 130
+#define UM_POLL_PERIOD_MS 100
+#define UM_TIMEOUT_MS 1000
+
+#define UM_CMD_POLL 0xf0
+
+static const struct rdtech_um_channel rdtech_default_channels[] = {
+	{ "V",  2, RDTECH_UM_UINT16, 0.01, 2, SR_MQ_VOLTAGE, SR_UNIT_VOLT },
+	{ "I",  4, RDTECH_UM_UINT16, 0.001, 3, SR_MQ_CURRENT, SR_UNIT_AMPERE },
+	{ "D+", 96, RDTECH_UM_UINT16, 0.01, 2, SR_MQ_VOLTAGE, SR_UNIT_VOLT },
+	{ "D-", 98, RDTECH_UM_UINT16, 0.01, 2, SR_MQ_VOLTAGE, SR_UNIT_VOLT },
+	{ "Temp", 10, RDTECH_UM_UINT16, 1.0, 0, SR_MQ_TEMPERATURE, SR_UNIT_CELSIUS },
+	/* Threshold-based recording (mWh) */
+	{ "Consumption", 106, RDTECH_UM_UINT32, 0.001, 3, 0, SR_UNIT_WATT_HOUR },
```

Comment:
I'd suggest to rename the temperature channel to `T`, like in other devices (e.g. in the `appa-55ii` driver).
The consumption channel should be renamed to `Wh` or `E` for energy (or `W` for work, but `W` would be ambiguous to Watt). I'd prefer `Wh` :)

Also the measured quantity is missing here. I mean there is no `SR_MQ_ENERGY` (or `SR_MQ_WORK`) in the `sr_mq` enum (`include/libsigrok/libsigrok.h`). Would be great, when you would add the missing mq to the enum (at the end of the enum to maintain API/ABI compability), to the `sr_key_info_mq` struct in `src/hwdriver.c` and to `bindings/cxx/enums.cpp`.

Same for the `rdtech_um25c_channels` down below.

Reply:
Thanks for the review! I think I prefer `E` to `Wh` for the measured energy consumption. I have updated the commit accordingly.

Reply:
Yeah, `E` or `Wh` doesn't really matter. I'm using `Wh` for detecting and naming energy channels in SmuView, but I'm going to add `E` to the detection part.

You missed `bindings/cxx/enums.cpp` when adding `SR_MQ_ENERGY`:

- At line 116:

        const Quantity Quantity::_ENERGY = Quantity(SR_MQ_ENERGY, "ENERGY");
- At line 149:

        const Quantity * const Quantity::ENERGY = &Quantity::_ENERGY;
- At line 183:

        {SR_MQ_ENERGY, Quantity::ENERGY},

Reply:
Isn't that file auto-generated? I did some grepping in the repo earlier to make sure I didn't miss anything and it didn't show up.

Reply:
Ups, I'm sorry. Of course, you are right!